models/base.py

```diff
@@ -268,6 +326,215 @@ def get_downstreams(self, session=None):
         """Get all data products that were created directly from this object (non-recursive)."""
         raise NotImplementedError('get_downstreams not implemented for this class')
 
+    def to_dict(self):
+        """Translate all the SQLAlchemy columns into a dictionary.
+
+        This can be used, e.g., to cache a row from DB to a file.
+        This will include foreign keys, which are not guaranteed
+        to remain the same when loading into a new database,
+        so all the relationships the object has should be
+        reconstructed manually when loading it from the dictionary.
+
+        This will not include any of the attributes of the object
+        that are not saved into the database, but those have to
+        be lazy loaded anyway, as they are not persisted.
+
+        Will convert non-standard data types:
+        UUID will be converted to string (using the .hex attribute).
+
+        To reload, use the from_dict() method:
+        reloaded_object = MyClass.from_dict( output_dict )
+        This will reconstruct the object, including the non-standard
+        data types like the UUID.
+        """
+        output = {}
+        for key in sa.inspect(self).mapper.columns.keys():
+            value = getattr(self, key)
+            if key == 'md5sum' and value is not None:
+                if isinstance(value, UUID):
+                    value = value.hex
+            if key == 'md5sum_extensions' and value is not None:
+                if isinstance(value, list):
+                    value = [v.hex if isinstance(v, UUID) else v for v in value]
+
+            if key == 'aper_rads' and isinstance(value, np.ndarray):
```

Comment:
I wonder if it might be worth keeping a list somewhere that lists all the keys that are array keys, and another list that are date/time keys.  Then you could just iterate through those lists here, and if we add new keys that fit those definitions, we just add them to the lists rather than adding to the if block.

Either way, we need to think about how we're going to make sure we (and anybody else who expands the code) remembers that something needs to get edited somewhere whenever defining a new database column that's either an array or a datetime.

Reply:
Maybe add an error in case a column has a forbidden type at the end of this call.